xmtp/src/networking.rs

```diff
@@ -0,0 +1,85 @@
+use std::collections::HashMap;
+
+use async_trait::async_trait;
+use xmtp_networking::grpc_api_helper::Subscription;
+use xmtp_proto::xmtp::message_api::v1::{Envelope, PagingInfo, PublishResponse, QueryResponse};
+
+#[async_trait]
+pub trait XmtpApiClient {
+    async fn publish(
+        &mut self,
+        token: String,
+        envelopes: Vec<Envelope>,
+        // TODO: use error enums
+    ) -> Result<PublishResponse, String>;
+
+    async fn query(
+        &mut self,
+        topic: String,
+        start_time: Option<u64>,
+        end_time: Option<u64>,
+        paging_info: Option<PagingInfo>,
+        // TODO: use error enums
+    ) -> Result<QueryResponse, String>;
+
+    // TODO: use error enums
+    async fn subscribe(&mut self, topics: Vec<String>) -> Result<Subscription, String>;
+}
+
+pub struct MockXmtpApiClient {
+    messages: HashMap<String, Vec<Envelope>>,
+}
+
+impl MockXmtpApiClient {
+    pub fn new() -> Self {
+        Self {
+            messages: HashMap::new(),
+        }
+    }
+}
+
+impl Default for MockXmtpApiClient {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[async_trait]
+impl XmtpApiClient for MockXmtpApiClient {
+    async fn publish(
+        &mut self,
+        token: String,
+        envelopes: Vec<Envelope>,
+    ) -> Result<PublishResponse, String> {
+        let mut existing: Vec<Envelope> = match self.messages.get(&token) {
```

Comment:
This is my half-finished garbage code. It shouldn't be looking up the messages by token, but going through each envelope and finding them by topic. You can see a correctly working version [here](https://github.com/xmtp/libxmtp/pull/118/files#diff-1c0e183cafb21ff8ca78b4d5a769750eb7df51979ca1f3316bd9a32786c9e645R54-R62).